### PR TITLE
Add Ubuntu 18.04 support

### PR DESCRIPTION
**Pull request by bastelfreak**
I'm currently implementing Ubuntu 18.04 support. Early feedback is highly appreciated :)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -43,6 +43,9 @@ group :system_tests do
   gem "beaker-pe",                                                               require: false
   gem "beaker-hostgenerator"
   gem "beaker-rspec"
+  gem 'rbnacl', '~> 4',                     :require => false if RUBY_VERSION >= '2.2.6'
```
Comment (HelenCampbell) [inline Gemfile]: What are you using these Gems for? The Gemfile will be overwritten by a pdk run so these changes would eventually get removed. If you want them permanent they'll have to be moved to .sync.yml.

File: manifests/mod/dav_svn.pp [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
   ::apache::mod { 'dav_svn': }
 
-  if $::osfamily == 'Debian' and ($::operatingsystemmajrelease != '6' and $::operatingsystemmajrelease != '10.04' and $::operatingsystemrelease != '10.04' and $::operatingsystemmajrelease != '16.04') {
+  if $::osfamily == 'Debian' and ($::operatingsystemmajrelease != '6' and !($::operatingsystemmajrelease in ['10.04', '16.04', '18.04'])) {
```
Comment (HelenCampbell) [inline manifests/mod/dav_svn.pp]: We have long dropped support for 10.04 so you can remove it here :)

File: spec/acceptance/mod_suphp_spec.rb [hunk 3]
```diff
@@ -1,6 +1,19 @@
...
 describe 'apache::mod::suphp class', if: (fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemmajrelease') != '16.04') do
+  if ['16.04', '18.04'].include?(fact('operatingsystemmajrelease'))
+    # this policy defaults to 101. it prevents newly installed services from starting
+    # it is useful for containers, it prevents new processes during 'docker build'
+    # but we actually want to test the services and this should not behave like docker
+    # but like a normal operating system
+
+    # without this apache fails to start -> installation of mod-php-something fails because it reloads apache to enable the module. WTF Ubuntu
+    # exit codes are documented at https://askubuntu.com/a/365912. Default for docker images is 101
+    shell("sed -i 's/^exit.*/exit 0/' /usr/sbin/policy-rc.d")
```
Comment (HelenCampbell) [inline spec/acceptance/mod_suphp_spec.rb]: Would it be worth adding this code into the spec_helper_acceptance since you reuse it a couple of times? Also gross to these bugs :|  I am running this PR through an ad-hoc pipeline and am hitting this issue on Ubuntu 16 as you seem to be.

File: spec/acceptance/mod_php_spec.rb [hunk 4]
```diff
@@ -120,18 +147,21 @@ class {'apache::mod::php':
...
+             else
+               "#{$mod_dir}/php5.conf"
+             end
+      describe file(path) do
+        it { is_expected.to contain '# somecontent' }
       end
     end
 
     context 'provide content and template config file' do
+      # does the following even makes sense? Why do we hardcode the php5 template?
```
Comment (HelenCampbell) [inline spec/acceptance/mod_php_spec.rb]: I'd agree with you, not sure there's much point to this test

File: Gemfile (hunk 1 above)
Comment (bastelfreak) [reply Gemfile]: The commit already has some information:
https://github.com/puppetlabs/puppetlabs-apache/pull/1809/commits/7703346bc27e5d264204a429a2a37041b6d4907d

File: spec/acceptance/mod_suphp_spec.rb (hunk 3 above)
Comment (bastelfreak) [reply spec/acceptance/mod_suphp_spec.rb]: Did you run the master branch and that failed, or my version?

File: manifests/mod/dav_svn.pp (hunk 2 above)
Comment (bastelfreak) [reply manifests/mod/dav_svn.pp]: removed it

File: Gemfile (hunk 1 above)
Comment (HelenCampbell) [reply Gemfile]: Thanks! We will look in to adding these to puppet-module-gems.

Comment (bastelfreak) [reply Gemfile]: We changed it to allow newer version of rbnacl as well: https://github.com/voxpupuli/modulesync_config/commit/58f32bf08917444081c84ffb9b58783dda4f31e6

File: spec/acceptance/mod_php_spec.rb [hunk 5]
```diff
@@ -4,6 +4,29 @@
...
+        # it is useful for containers, it prevents new processes during 'docker build'
+        # but we actually want to test the services and this should not behave like docker
+        # but like a normal operating system
+
+        # without this apache fails to start -> installation of mod-php-something fails because it reloads apache to enable the module. WTF Ubuntu
+        # exit codes are documented at https://askubuntu.com/a/365912. Default for docker images is 101
+        shell("sed -i 's/^exit.*/exit 0/' /usr/sbin/policy-rc.d")
+      end
+      if fact('operatingsystemmajrelease') == '18.04'
+        # Again Ubuntu. W T F.
```
Comment (rcousens) [inline spec/acceptance/mod_php_spec.rb]: I'd just like to point out this is an upstream Debian bug, so your 'WTF Ubuntu' is misdirected and misplaced. Regardless your personal feelings and reactions should most likely never appear in public code in the first place even if misguided.

File: spec/acceptance/mod_suphp_spec.rb (hunk 3 above)
Comment (HelenCampbell) [reply spec/acceptance/mod_suphp_spec.rb]: Your version. I'll re-run again this morning with your recent changes :)

File: spec/acceptance/mod_php_spec.rb (hunk 5 above)
Comment (bastelfreak) [reply spec/acceptance/mod_php_spec.rb]: sorry, that was written after hours of personal frustration, I removed it

Comment (rcousens) [reply spec/acceptance/mod_php_spec.rb]: I know the feeling, thanks for cleaning it up!

**Comment by tphoney** [comment]
Having a look, what a PR 😮

File: .sync.yml [hunk 6]
```diff
@@ -5,6 +5,8 @@
   docker_sets:
     - set: docker/centos-7
     - set: docker/ubuntu-14.04
+    - set: docker/ubuntu-16.04
```
Comment (HelenCampbell) [inline .sync.yml]: Can we get both Ubuntu16&18 removed from here and the travis file.

File: .travis.yml [hunk 7]
```diff
@@ -22,17 +22,33 @@ matrix:
   fast_finish: true
   include:
     -
-      bundler_args: 
+      bundler_args:
       dist: trusty
-      env: PUPPET_INSTALL_TYPE=agent BEAKER_debug=true BEAKER_PUPPET_COLLECTION=puppet5 BEAKER_set=docker/centos-7
+      env: PUPPET_INSTALL_TYPE=agent BEAKER_debug=true BEAKER_PUPPET_COLLECTION=puppet5 BEAKER_setfile=centos7-64{hypervisor=docker}
```
Comment (HelenCampbell) [inline .travis.yml]: Can we get this change reverted back to BEAKER_set=docker/centos-7 and BEAKER_set=docker/ubuntu-14.04 please? (Work is ongoing to implement this functionality.)

Comment (bastelfreak) [reply .travis.yml]: done

File: .sync.yml (hunk 6 above)
Comment (bastelfreak) [reply .sync.yml]: done

File: Gemfile (hunk 1 above)
Comment (HelenCampbell) [reply Gemfile]: Can you remove these new Gems for merge? We will be adding them to the puppet-module-gems in due time and these would conflict.

File: spec/spec_helper_local.rb [hunk 8]
```diff
@@ -35,5 +35,5 @@
     }
   end
 
-  it { is_expected.to compile.with_all_deps }
+  # it { is_expected.to compile.with_all_deps }
```
Comment (HelenCampbell) [inline spec/spec_helper_local.rb]: Is there any reason this is commented out?

Comment (bastelfreak) [reply spec/spec_helper_local.rb]: It didn't work in combination with rspec-puppet-facts and also felt redundant. All the mod files should already include `it { is_expected.to compile.with_all_deps }`

File: spec/acceptance/mod_suphp_spec.rb (hunk 3 above)
Comment (HelenCampbell) [reply spec/acceptance/mod_suphp_spec.rb]: So I've found locally that my tests will run if I change to 
`  before :all do
    on(host, "if [ -a '/usr/sbin/policy-rc.d' ]; then sed -i 's/^exit.*/exit 0/' /usr/sbin/policy-rc.d; fi") if ['16.04', '18.04'].include?(fact('operatingsystemmajrelease'))
  end`
instead of doing it in the describe.

Comment (HelenCampbell) [reply spec/acceptance/mod_suphp_spec.rb]: Ew that didn't format well.. To summarise I just moved it into a 'before :all' clause. That seems to be why I kept hitting into that issue where the tests bombed out instead of running.

File: manifests/mod/dav_svn.pp [hunk 9]
```diff
@@ -13,7 +13,7 @@
 
   ::apache::mod { 'dav_svn': }
 
-  if $::osfamily == 'Debian' and ($::operatingsystemmajrelease != '6' and $::operatingsystemmajrelease != '10.04' and $::operatingsystemrelease != '10.04' and $::operatingsystemmajrelease != '16.04') {
+  if $::osfamily == 'Debian' and ($::operatingsystemmajrelease != '6' and !($::operatingsystemmajrelease in ['16.04', '18.04'])) {
```
Comment (ekohl) [inline manifests/mod/dav_svn.pp]: The `$::operatingsystemmajrelease != '6'` looks redundant. I hope Debian 6 isn't supported. It's also pretty hard to parse. So the result is that for Ubuntu 14.04 it's `undef` but everything else has `dav_svn_authz_svn.load`?

File: manifests/mod/security.pp [hunk 10]
```diff
@@ -128,6 +128,8 @@
     }
   }
 
-  unless $::operatingsystem == 'SLES' { apache::security::rule_link { $activated_rules: } }
+  # the rule setup on ubuntu 18.04 is completely different,
+  # the default rules used within this module aren't available
+  unless $::operatingsystem == 'SLES' or $facts['os']['release']['major'] == '18.04' { apache::security::rule_link { $activated_rules: } }
```
Comment (ekohl) [inline manifests/mod/security.pp]: Might be a bit more readable to put the body on a new line. It's very long.

File: manifests/mod/suphp.pp [hunk 11]
```diff
@@ -1,5 +1,8 @@
 class apache::mod::suphp (
 ){
+  if ($facts['os']['name'] == 'Ubuntu' and versioncmp($facts['os']['release']['major'], '18.04') >= 0) {
+    fail('mod_fastcgi is no longer supported on Ubuntu 18.04 and above. Please use mod_proxy_fcgi')
```
Comment (ekohl) [inline manifests/mod/suphp.pp]: This message doesn't make sense to me. Does suphp depend on `mod_fastcgi`? Copy paste error?

File: manifests/params.pp [hunk 12]
```diff
@@ -263,6 +263,33 @@
         'wsgi'                  => 'libapache2-mod-wsgi',
         'xsendfile'             => 'libapache2-mod-xsendfile',
       }
+    } elsif $facts['os']['release']['major'] == '18.04' {
+      # major.minor version used since Debian stretch and Ubuntu Xenial
+      $php_version = '7.2' # different to Ubuntu 18.04
```
Comment (ekohl) [inline manifests/params.pp]: Did you mean 16.04 here?

File: metadata.json [hunk 13]
```diff
@@ -21,15 +21,13 @@
     {
       "operatingsystem": "RedHat",
       "operatingsystemrelease": [
-        "5",
```
Comment (ekohl) [inline metadata.json]: Is this the right PR to drop these?

File: spec/acceptance/mod_passenger_spec.rb [hunk 14]
```diff
@@ -10,10 +10,7 @@
...
     when 'Ubuntu'
       case fact('lsbdistrelease')
-      when '14.04'
-        passenger_root = '/usr/lib/ruby/vendor_ruby/phusion_passenger/locations.ini'
-        passenger_default_ruby = '/usr/bin/ruby'
-      when '16.04'
+      when '14.04', '16.04', '18.04'
         passenger_root = '/usr/lib/ruby/vendor_ruby/phusion_passenger/locations.ini'
         passenger_default_ruby = '/usr/bin/ruby'
       else
```
Comment (ekohl) [inline spec/acceptance/mod_passenger_spec.rb]: These versions are no longer supported so you can simplify the whole block. Perhaps Debian deserves the same treatment.

File: spec/acceptance/mod_passenger_spec.rb [hunk 15]
```diff
@@ -113,12 +110,28 @@ class { 'apache::mod::passenger':
...
-          class { 'apache': }
-          class { 'apache::mod::passenger': }
-          #{pp_rackapp}
-      MANIFEST
+      # We need to set passenger_instance_registry_dir on every sane distro
+      # with systemd. Systemd can force processes into a seperate/private
+      # tmpdir. This is the default for apache on Ubuntu 18.04. As a result,
+      # passenger CLI tools can't find the config/socket, which defaults to /tmp
+      # we enable it for ubuntu 16.04/18.04, centos7 and debian 9
+      pp = if ['7', '9', '16.04', '18.04'].include?(fact('operatingsystemmajrelease'))
```
Comment (ekohl) [inline spec/acceptance/mod_passenger_spec.rb]: These magic numbers feel like a difficult to maintain/easy to break setup. Perhaps it should check for `fact('service_provider') == 'systemd'` instead? You could even do it in a pure puppet block:
```puppet
$passenger_instance_registry_dir = $fact['service_provider'] == 'systemd' ?
  true    => '/var/run',
  default => undef,
}
class { 'apache': }
class { 'apache::mod::passenger':
  passenger_instance_registry_dir => $passenger_instance_registry_dir,
}
```

File: spec/acceptance/mod_passenger_spec.rb [hunk 16]
```diff
@@ -137,10 +150,7 @@ class { 'apache::mod::passenger': }
...
         when 'Ubuntu'
           case fact('lsbdistrelease')
-          when '14.04'
-            it { is_expected.to contain %(PassengerDefaultRuby "#{passenger_default_ruby}") }
-            it { is_expected.not_to contain '/PassengerRuby/' }
-          when '16.04'
+          when '14.04', '16.04', '18.04'
             it { is_expected.to contain %(PassengerDefaultRuby "#{passenger_default_ruby}") }
             it { is_expected.not_to contain '/PassengerRuby/' }
           else
```
Comment (ekohl) [inline spec/acceptance/mod_passenger_spec.rb]: The else block is unsupported as well, same as earlier.

File: spec/acceptance/mod_passenger_spec.rb [hunk 17]
```diff
@@ -171,10 +181,9 @@ class { 'apache::mod::passenger': }
       expected_one = [%r{Apache processes}, %r{Nginx processes}, %r{Passenger processes}]
       # passenger-memory-stats output on newer Debian/Ubuntu verions do not contain
       # these two lines
-      unless (fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemrelease') == '14.04') ||
-             (fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemrelease') == '16.04') ||
+      unless (fact('operatingsystem') == 'Ubuntu' && ['14.04', '16.04', '18.04'].include?(fact('operatingsystemrelease'))) ||
```
Comment (ekohl) [inline spec/acceptance/mod_passenger_spec.rb]: Isn't this basically every supported Ubuntu version? I'd even thing Debian 9 would fail here so I think this whole block can be removed if it's no longer included.

File: spec/acceptance/mod_passenger_spec.rb [hunk 18]
```diff
@@ -193,27 +202,26 @@ class { 'apache::mod::passenger': }
...
         it 'outputs status via passenger-status #General information' do
-          shell('PATH=/usr/bin:$PATH /usr/sbin/passenger-status') do |r|
+          shell('PATH=/usr/bin:$PATH PASSENGER_INSTANCE_REGISTRY_DIR=/var/run /usr/sbin/passenger-status') do |r|
             # spacing may vary
             expect(r.stdout).to match(%r{[\-]+ General information [\-]+})
           end
         end
-        expected_two = if fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemrelease') == '14.04' ||
-                          (fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemrelease') == '16.04') ||
+        expected_two = if fact('operatingsystem') == 'Ubuntu' && ['14.04', '16.04', '18.04'].include?(fact('operatingsystemrelease')) ||
```
Comment (ekohl) [inline spec/acceptance/mod_passenger_spec.rb]: I think you can drop the Ubuntu version checks here as well.

File: spec/acceptance/mod_suphp_spec.rb [hunk 19]
```diff
@@ -1,6 +1,19 @@
...
+  end
+  # mod php isn't supported for ubuntu 1.04
+  next if fact('operatingsystemmajrelease') == '18.04'
```
Comment (ekohl) [inline spec/acceptance/mod_suphp_spec.rb]: It's better to do this in the describe like 16.04.

File: spec/acceptance/mod_suphp_spec.rb [hunk 20]
```diff
@@ -1,6 +1,19 @@
 require 'spec_helper_acceptance'
 
+# why is that not executed on 16.04
```
Comment (ekohl) [inline spec/acceptance/mod_suphp_spec.rb]: And why is it only executed on Ubuntu setups? Do other distros not ship it?

File: spec/acceptance/vhost_spec.rb [hunk 21]
```diff
@@ -964,7 +964,7 @@ class { 'apache': }
     describe file($ports_file) do
       it { is_expected.to be_file }
       if fact('osfamily') == 'RedHat' && fact('operatingsystemmajrelease') == '7' ||
-         fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemrelease') =~ %r{(14\.04|16\.04)} ||
+         fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemrelease') =~ %r{(14\.04|16\.04|18\.04)} ||
```
Comment (ekohl) [inline spec/acceptance/vhost_spec.rb]: I think this is all supported Ubuntu versions.

File: manifests/params.pp (hunk 12 above)
Comment (bastelfreak) [reply manifests/params.pp]: fixed

File: spec/acceptance/mod_suphp_spec.rb (hunk 3 above)
Comment (bastelfreak) [reply spec/acceptance/mod_suphp_spec.rb]: I added it like this:

```patch
 spec/acceptance/mod_suphp_spec.rb | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/spec/acceptance/mod_suphp_spec.rb b/spec/acceptance/mod_suphp_spec.rb
index dbf3a28c..4a1388b8 100644
--- a/spec/acceptance/mod_suphp_spec.rb
+++ b/spec/acceptance/mod_suphp_spec.rb
@@ -1,8 +1,8 @@
 require 'spec_helper_acceptance'
 
 # why is that not executed on 16.04
-describe 'apache::mod::suphp class', if: (fact('operatingsystem') == 'Ubuntu' && fact('operatingsystemmajrelease') != '16.04') do
-  if ['16.04', '18.04'].include?(fact('operatingsystemmajrelease'))
+describe 'apache::mod::suphp class' do
+  before :all do
     # this policy defaults to 101. it prevents newly installed services from starting
     # it is useful for containers, it prevents new processes during 'docker build'
     # but we actually want to test the services and this should not behave like docker
@@ -10,8 +10,9 @@ describe 'apache::mod::suphp class', if: (fact('operatingsystem') == 'Ubuntu' &&
 
     # without this apache fails to start -> installation of mod-php-something fails because it reloads apache to enable the module. WTF Ubuntu
     # exit codes are documented at https://askubuntu.com/a/365912. Default for docker images is 101
-    shell("sed -i 's/^exit.*/exit 0/' /usr/sbin/policy-rc.d")
+    on(host, "if [ -a '/usr/sbin/policy-rc.d' ]; then sed -i 's/^exit.*/exit 0/' /usr/sbin/policy-rc.d; fi") if ['16.04', '18.04'].include?(fact('operatingsystemmajrelease'))
   end
+
   # mod php isn't supported for ubuntu 1.04
   next if fact('operatingsystemmajrelease') == '18.04'
   context 'default suphp config' do
```

but I get this error:

```
Failures:

  1) apache::mod::suphp class default suphp config succeeds in puppeting suphp
     Failure/Error: on(host, "if [ -a '/usr/sbin/policy-rc.d' ]; then sed -i 's/^exit.*/exit 0/' /usr/sbin/policy-rc.d; fi") if ['16.04', '18.04'].include?(fact('operatingsystemmajrelease'))
     NoMethodError:
       private method `exec' called for Host "":Serverspec::Type::Host
```

File: manifests/mod/dav_svn.pp (hunk 9 above)
Comment (bastelfreak) [reply manifests/mod/dav_svn.pp]: cleaned it up

File: manifests/mod/security.pp (hunk 10 above)
Comment (bastelfreak) [reply manifests/mod/security.pp]: fixed

File: metadata.json (hunk 13 above)
Comment (bastelfreak) [reply metadata.json]: cherry-picked into https://github.com/puppetlabs/puppetlabs-apache/pull/1824

File: spec/acceptance/mod_passenger_spec.rb (hunk 14 above)
Comment (bastelfreak) [reply spec/acceptance/mod_passenger_spec.rb]: cleaned it up

File: spec/acceptance/mod_passenger_spec.rb (hunk 16 above)
Comment (bastelfreak) [reply spec/acceptance/mod_passenger_spec.rb]: fixed

File: spec/acceptance/mod_passenger_spec.rb (hunk 17 above)
Comment (bastelfreak) [reply spec/acceptance/mod_passenger_spec.rb]: fixed

File: spec/acceptance/mod_passenger_spec.rb (hunk 18 above)
Comment (bastelfreak) [reply spec/acceptance/mod_passenger_spec.rb]: done

File: spec/acceptance/vhost_spec.rb (hunk 21 above)
Comment (bastelfreak) [reply spec/acceptance/vhost_spec.rb]: fixed

**Comment by mmoll** [comment]
this needs a rebase...
without having a closer look: what's backwards-imcompatible here?

**Comment by david22swan** [comment]
@bastelfreak Closing as a pr finishing this work has been merged in:
https://github.com/puppetlabs/puppetlabs-apache/pull/1850
Thanks for the great work you put into it though
